### PR TITLE
Make the implicitly exported rule consistent between checker and binder

### LINKED ISSUE #54342: Default imported function merged with namespace cannot be called (has no call signatures)
# Bug Report

### 🔎 Search Terms

import default function namespace call signatures

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about modules

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBFApmANgQwMaIJIDsBiwKicAZlBCHAERKqaIC0wujpRi1A3ALABQ-Ouix5CxABQBvMhwDOALhoAqAHQgAJtQA0ZCiEXU02uDAgGARtTgBfAJS8+QA)

### 💻 Code

```ts
import replaceInFile from "replace-in-file";

replaceInFile({ files: "*.md", from: "a", to: "b" });
```

### 🙁 Actual behavior

```plaintext
This expression is not callable.
  Type 'typeof replaceInFile' has no call signatures. (ts2349)
```

### 🙂 Expected behavior

`replaceInFile` should be callable. It's a `function` and even is known to have function properties such as `.apply`, `.bind`, and `.call`.

If there actually is something off about the code with respect to modules and importing, then the error should be made more clear. This is very confusing. 😞 

Seems to be a similar issue to #50175?

### PR DESCRIPTION
**Pull request by gabritto**
Fixes #54342.

In the binder, when deciding if a declaration should be treated as exported or not, [we consult the declaration container's flags to see if it is an "export context"](https://github.com/microsoft/TypeScript/blob/972299ae347bd29cfb3be2de1a7f2ee24e0781cd/src/compiler/binder.ts#L902), meaning an ambient context where declarations are implicitly exported. In the checker, however, [we treat everything in an ambient context as exported](https://github.dev/microsoft/TypeScript/blob/d79ec186d6a4e39f57af6143761d453466a32e0c/src/compiler/checker.ts#L39027-L39030), regardless of whether we're in an ambient context where declarations are implicitly exported, resulting in us incorrectly not erroring on the `.d.ts` file present in #54342.

### CODE REVIEW

**Comment by gabritto** [comment]
@typescript-bot run DT

**Comment by gabritto** [comment]
Well, I guess the bot still isn't working, so we'll have to wait to see what this PR breaks in DT.

**Comment by gabritto** [comment]
@typescript-bot run DT

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

pretty please

**Comment by jakebailey** [comment]
I am moderately suspicious that the user tests didn't net anything, but maybe that's to be expected if this only happened in ambient contexts. (I guess top100 is still running)

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -455,7 +456,8 @@ function getModuleInstanceStateForAliasTarget(specifier: ExportSpecifier, visite
     return ModuleInstanceState.Instantiated; // Couldn't locate, assume could refer to a value
 }
 
-const enum ContainerFlags {
+/** @internal */
+export const enum ContainerFlags {
```
Comment (jakebailey) [inline src/compiler/binder.ts:460]: I'm not experienced enough to know the right way to do this, but it does feel a little suspicious to export... I feel like I've seen a function that does something similar in the checker but I can't recall the name.

Comment (gabritto) [reply src/compiler/binder.ts:460]: I wanted to make sure the `container` used to check the implicit export rules is precisely the same the binder uses, and the binder uses those flags for that... I saw a `getContainerNode` in services/utilities, and `getEnclosingBlockScopeContainer` in compiler/utilities, but those are different.

**Review by andrewbranch — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Because the `export default` feels like a red herring, I'll just mention that

```ts
declare module 'replace-in-file' {
    export function replaceInFile(config: unknown): Promise<unknown[]>;
    export {};

    namespace replaceInFile {
        export function sync(config: unknown): unknown[];
  }
}
```

introduces the same issue.

In an ambient module, adding an `export { ... }` or a similar construct like `export default ...` implicitly changes whether all declarations are automatically exported.

TypeScript now recognizes these unfortunately confusing semantics more consistently, and issues an error on the fact that all declarations of `replaceInFile` need to agree in their modifiers.

In the case where there is no `export {}` or `export default`, there is no error because everything is implicitly exported anyway.

```ts
declare module 'replace-in-file' {
    export function replaceInFile(config: unknown): Promise<unknown[]>;

    namespace replaceInFile {
        export function sync(config: unknown): unknown[];
  }
}
```

**Pull request merged**